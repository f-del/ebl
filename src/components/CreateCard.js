import React, { useState } from "react";
import PropTypes from "prop-types";

const MODE_READ = 1;
const MODE_FORM = 2;

function CreateCard({ onValidate }) {
  let [mode, setMode] = useState(MODE_READ);
  let [value, setValue] = useState("");

  const handleKeys = e => {
    const isEnterKey = e.which === 13;
    if (e.which === 27) {
      setValue("");
      setMode(MODE_READ);
    } else {
      if (value !== "" && isEnterKey) {
        onValidate(value);
        setValue("");
        setMode(MODE_READ);
      }
    }
    if (isEnterKey) e.preventDefault();
  };

  const displayForm = () => (
    <textarea
      onKeyDown={handleKeys}
      onChange={e => setValue(e.target.value)}
      rows="8"
      cols="17"
      autoFocus={true}
    />
  );
  const displayButton = () => (
    <button onClick={() => setMode(MODE_FORM)}>Create new task</button>
  );

  return (
    <div className={`create-card${mode === MODE_FORM ? "--actif" : ""}`}>
      {mode === MODE_FORM && displayForm()}
      {mode === MODE_READ && displayButton()}
    </div>
  );
}

CreateCard.propTypes = {
  onValidate: PropTypes.func.isRequired
};

export default CreateCard;
