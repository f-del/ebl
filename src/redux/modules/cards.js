import update from "immutability-helper";

const CREATE = "CARD/CREATE";
const START = "CARD/START";

const card = title => ({
  Title: title,
  Status: cardStatus.Todo,
  Type: cardType.Task
});

export const cardStatus = {
  Todo: "TODO",
  Inprogress: "INPROGRESS"
};

export const cardType = {
  Task: "TASK"
};

export const createCardSuccess = (id, card) => ({
  type: CREATE,
  payload: { Id: id, ...card }
});

export const createCard = title => {
  return (dispatch, getState, { api }) => {
    const newCard = card(title);
    return api.Tasks.Post(newCard).then(res => {
      console.log("post done,  new Id =" + res.Id);
      dispatch(createCardSuccess(res.Id, newCard));
    });
  };
};

export const startCard = id => {
  if (id === undefined) throw new Error("Id arg is mandatory");
  return {
    type: START,
    payload: {
      Id: id
    }
  };
};

const initialState = {
  list: []
};

export default function(state = initialState, action) {
  console.debug(action);
  switch (action.type) {
    case CREATE:
      return { ...state, list: [...state.list, action.payload] };
    case START:
      let idx = state.list.findIndex(c => c.Id === action.payload.Id);

      var mergeObject = {};
      mergeObject[idx] = { $merge: { Status: cardStatus.Inprogress } };

      return { ...state, list: update(state.list, mergeObject) };

    default:
      return state;
  }
}

export function getAllCards(state) {
  if (state === undefined) throw new Error("State arg is mandatory");
  return state.cards !== undefined ? state.cards.list : [];
}

export function getAllCardsInProgess(state) {
  return getAllCards(state).filter(isCardStatusInProgess);
}
export function getAllCardsTodo(state) {
  return getAllCards(state).filter(isCardStatusToDo);
}

/* Functionnal helper */
function isCardStatusInProgess(card) {
  return card.Status === cardStatus.Inprogress;
}
function isCardStatusToDo(card) {
  return card.Status === cardStatus.Todo;
}
