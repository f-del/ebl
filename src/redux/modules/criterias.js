export const criteriaType = {
  BASIC: "BASIC",
  DEV: "DEV"
};

export function create(id, value) {
  return { Id: id, Value: value };
}

const initialState = {
  list: {
    BASIC: [
      {
        Id: "IdCritBas1",
        Value: false
      }
    ],
    DEV: [
      {
        Id: "IdCritDev1",
        Text: "Development",
        Value: false
      },
      {
        Id: "IdCritDev2",
        Text: "Unit testing",
        Value: false
      }
    ]
  }
};

export default function(state = initialState, action) {
  return state;
}

export function getCriteria(state, type) {
  if (type === undefined) throw new Error("Type argument is mandatory");
  return getState(state).list[type] || [];
}
function getState(state) {
  return state.criterias || {};
}
