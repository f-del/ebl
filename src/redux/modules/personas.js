/***
 *    ██████╗ ███████╗██████╗ ██╗   ██╗ ██████╗███████╗██████╗ ███████╗
 *    ██╔══██╗██╔════╝██╔══██╗██║   ██║██╔════╝██╔════╝██╔══██╗██╔════╝
 *    ██████╔╝█████╗  ██║  ██║██║   ██║██║     █████╗  ██████╔╝███████╗
 *    ██╔══██╗██╔══╝  ██║  ██║██║   ██║██║     ██╔══╝  ██╔══██╗╚════██║
 *    ██║  ██║███████╗██████╔╝╚██████╔╝╚██████╗███████╗██║  ██║███████║
 *    ╚═╝  ╚═╝╚══════╝╚═════╝  ╚═════╝  ╚═════╝╚══════╝╚═╝  ╚═╝╚══════╝
 *
 */

const initialState = {
  list: {
    e6wTJIBsAw5puwwrNEsR: {
      Name: "Feature Team",
      Details: "Meta persona, representing all collaborators working with Paul",
      Needs: [
        "To understand WHAT the Product Owner wants to be able to design collaborativly the HOW",
        "To manage easly the list of things to do to finish the most quickly the User Story"
      ],
      Picture: ""
    },
    uOCF2GLFiJl6fvxdlq4b: {
      Details: "Product Owner for a new product, work in a startup",
      Name: "Paul Deligny",
      Needs: [
        "To have a clear understanding about the product vision & objectives",
        "To write real User Stories, ie have a Customer Driven thinking",
        "To share at the right time and with the rigth level of information the User Stories with his team",
        "To have a clear vision on the job he'll have to do AND the job already or currently done"
      ],
      Picture: ""
    }
  }
};

export default function(state = initialState, action) {
  return state;
}

/***
 *    ███████╗███████╗██╗     ███████╗ ██████╗████████╗ ██████╗ ██████╗ ███████╗
 *    ██╔════╝██╔════╝██║     ██╔════╝██╔════╝╚══██╔══╝██╔═══██╗██╔══██╗██╔════╝
 *    ███████╗█████╗  ██║     █████╗  ██║        ██║   ██║   ██║██████╔╝███████╗
 *    ╚════██║██╔══╝  ██║     ██╔══╝  ██║        ██║   ██║   ██║██╔══██╗╚════██║
 *    ███████║███████╗███████╗███████╗╚██████╗   ██║   ╚██████╔╝██║  ██║███████║
 *    ╚══════╝╚══════╝╚══════╝╚══════╝ ╚═════╝   ╚═╝    ╚═════╝ ╚═╝  ╚═╝╚══════╝
 *
 */
export function getAllPersonas(state) {
  return mapToArrayPersona(getPersonasList(state)) || [];
}

function getPersonasList(state) {
  return getPersonas(validateState(state)).list || {};
}

/* Functionnal helper */
function validateState(state) {
  if (state === undefined) throw new Error("State arg is mandatory");
  return state;
}
function getPersonas(state) {
  return state.personas || {};
}

function mapToArrayPersona(list) {
  return Array.from(Object.keys(list), key => ({
    ...getPersonaId(key),
    ...list[key]
  }));
}
function getPersonaId(key) {
  return {
    Id: key
  };
}
