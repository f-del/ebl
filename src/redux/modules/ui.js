const SELECT_HYPOTHESIS = "UI/SELECT/HYPOTHESIS";
const SELECT_USERSTORY = "UI/SELECT/USER-STORY";

export const selectHypothesis = hypothesis => ({
  type: SELECT_HYPOTHESIS,
  payload: {
    Id: hypothesis.Id
  }
});

export const selectUserStory = userstory => ({
  type: SELECT_USERSTORY,
  payload: {
    Id: userstory.Id
  }
});

const initialState = {
  currentHypothesis: undefined,
  currentUserStory: undefined
};

export default function uiReducer(state = initialState, action) {
  switch (action.type) {
    case SELECT_HYPOTHESIS: {
      // currently Reducer handle Ui logic -> deselect current US on Hypothesis change
      return {
        ...state,
        currentHypothesis: action.payload.Id,
        currentUserStory: undefined
      };
    }
    case SELECT_USERSTORY: {
      return { ...state, currentUserStory: action.payload.Id };
    }

    default:
      return state;
  }
}

export function getHypothesisSelected(state) {
  return state.ui.currentHypothesis;
}
export function getUserStorySelected(state) {
  return state.ui.currentUserStory;
}
