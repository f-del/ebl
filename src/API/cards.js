import { cardStatus } from "../redux/modules/cards";

const mapping = card => {
  card = card._firestore !== undefined ? card.data() : card;
  return {
    ...card,
    Status:
      card.Status === "INPROGRESS" ? cardStatus.INPROGRESS : cardStatus.TODO
  };
};

function post(db) {
  return card => {
    return new Promise((resolve, reject) => {
      db.collection("cards")
        .add(card)
        .then(ref => {
          console.log("Document successfully written!");
          resolve({ id: ref.id });
        })
        .catch(error => {
          console.error("Error writing document: ", error);
          //dispatch(errorSave())
          reject(error);
        });
    });
  };
}

function get(db) {
  return type => {
    return new Promise((resolve, reject) => {
      db.collection("cards")
        .where("Type", "==", type)
        .get()
        .then(function(querySnapshot) {
          let cards = [];

          querySnapshot.forEach(function(doc) {
            let env = mapping(doc);

            cards.push(env);
          });

          resolve(cards);
        });
    });
  };
}

export default db => {
  return { Post: post(db), Get: get(db) };
};
