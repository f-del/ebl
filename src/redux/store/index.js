import cards from "../modules/cards";
import { combineReducers } from "redux";

const rootReducer = combineReducers({ cards });
export default rootReducer;
