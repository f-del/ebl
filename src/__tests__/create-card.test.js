import React from "react";
import { shallow, configure, mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

import { createStore, applyMiddleware } from "redux";
import { Provider } from "react-redux";
import thunk from "redux-thunk";
import reducer from "../redux/store/index";

import * as cardsSelector from "../redux/modules/cards";
import CreateCard from "../components/CreateCard";
import CreateTaskCard from "../containers/createTaskCard";

configure({ adapter: new Adapter() });

let store;
const onValidate = jest.fn();

describe("Componant tests", () => {
  function expectInitState(wrapper) {
    expect(wrapper.find("button").length).toBe(1);
    expect(wrapper.find("button").text()).toBe("Create new task");
    expect(wrapper.find("input").length).toBe(0);
  }

  test("Initial state", () => {
    const wrapper = shallow(<CreateCard onValidate={onValidate} />);

    expectInitState(wrapper);
    //expect(wrapper.state("mode")).toBe(1);

    wrapper.unmount();
  });

  test("Click on create new task and cancel by ESCAPE", () => {
    const wrapper = shallow(<CreateCard onValidate={onValidate} />);

    expectInitState(wrapper);
    expect(wrapper.find("button").simulate("click"));

    //expect(wrapper.state("mode")).toBe(2);
    expect(wrapper.find("button").length).toBe(0);

    const input = wrapper.find("textarea");

    expect(input.length).toBe(1);

    const text = "Unit test task";
    [...text].forEach((c, i) => {
      input.simulate("keyDown", { which: c.charCodeAt(0) });
    });

    input.simulate("keyDown", { which: 27 });

    expect(wrapper.find("button").length).toBe(1);
    //expect(wrapper.state("mode")).toBe(1);
    wrapper.unmount();
  });

  test("Click on create new task and validate by ENTER", () => {
    const wrapper = mount(<CreateCard onValidate={onValidate} />);

    const preventDefault = jest.fn();
    //wrapper.setProps({ onValidate });

    expectInitState(wrapper);
    expect(wrapper.find("button").simulate("click"));

    expect(wrapper.find("button").length).toBe(0);

    const input = wrapper.find("textarea");
    expect(input.prop("autoFocus")).toBe(true);

    expect(input.length).toBe(1);
    // TODO : expect focus ?
    //expect(document.activeElement).toEqual(input.getDOMNode());

    const text = "Unit test task";
    // [...text].forEach((c, i) => {
    //   //input.simulate("keyDown", { which: c.charCodeAt(0) });
    // });
    input.simulate("change", { target: { value: text } });

    input.simulate("keyDown", { which: 13, preventDefault });
    expect(preventDefault.mock.calls.length).toBe(1);
    expect(onValidate.mock.calls.length).toBe(1);
    expect(onValidate.mock.calls[0][0]).toEqual(text);
    expectInitState(wrapper);

    wrapper.unmount();
  });
});

describe("Containers", () => {
  beforeEach(() => {
    jest.resetAllMocks();
    store = createStore(reducer, applyMiddleware(thunk));
  });

  test("createTask", () => {
    const text = "Unit test task";
    const preventDefault = jest.fn();
    cardsSelector.createCard = jest.fn().mockImplementationOnce(id => {
      return dispatch => {
        return id;
      };
    });

    const wrapper = mount(
      <Provider store={store}>
        <CreateTaskCard />
      </Provider>
    );
    expect(wrapper.props("onValidate")).toBeDefined();

    expect(wrapper.find("button").simulate("click"));
    const input = wrapper.find("textarea");
    expect(input.length).toBe(1);
    input.simulate("change", { target: { value: text } });
    input.simulate("keyDown", { which: 13, preventDefault });
    expect(cardsSelector.createCard.mock.calls.length).toBe(1);
    expect(cardsSelector.createCard.mock.calls[0][0]).toBe(text);

    wrapper.unmount();
  });
});
