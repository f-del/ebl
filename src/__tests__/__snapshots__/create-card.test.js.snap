// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Componant tests Click on create new task and cancel by ESCAPE 1`] = `
<div
  className="create-card--actif"
>
  <textarea
    autoFocus={true}
    cols="17"
    onChange={[Function]}
    onKeyDown={[Function]}
    rows="8"
  />
</div>
`;

exports[`Componant tests Click on create new task and validate by ENTER 1`] = `
<CreateCard
  onValidate={
    [MockFunction] {
      "calls": Array [
        Array [
          "Unit test task",
        ],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": undefined,
        },
      ],
    }
  }
>
  <div
    className="create-card"
  >
    <button
      onClick={[Function]}
    >
      Create new task
    </button>
  </div>
</CreateCard>
`;

exports[`Componant tests Initial state 1`] = `
<div
  className="create-card"
>
  <button
    onClick={[Function]}
  >
    Create new task
  </button>
</div>
`;

exports[`Containers createTask 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(CreateCard)>
    <CreateCard
      onValidate={[Function]}
    >
      <div
        className="create-card"
      >
        <button
          onClick={[Function]}
        >
          Create new task
        </button>
      </div>
    </CreateCard>
  </Connect(CreateCard)>
</Provider>
`;
